lightning-persister/src/lib.rs

```diff
@@ -0,0 +1,263 @@
+	// writes are queued to a separate writer thread. This would improve performance
+	// since syncing to disk can sometimes take 100s of milliseconds.
+	fn write_channel_data(&self, funding_txo: OutPoint, monitor: &dyn DiskWriteable) -> std::io::Result<()> {
+		fs::create_dir_all(&self.path_to_channel_data)?;
+		// Do a crazy dance with lots of fsync()s to be overly cautious here...
+		// We never want to end up in a state where we've lost the old data, or end up using the
+		// old data on power loss after we've returned.
+		// The way to atomically write a file on Unix platforms is:
+		// open(tmpname), write(tmpfile), fsync(tmpfile), close(tmpfile), rename(), fsync(dir)
+		let filename = self.get_full_filepath(funding_txo);
+		let tmp_filename = filename.clone() + ".tmp";
+
+		{
+			// Note that going by rust-lang/rust@d602a6b, on MacOS it is only safe to use
+			// rust stdlib 1.36 or higher.
+			let mut f = fs::File::create(&tmp_filename)?;
+			monitor.write(&mut f)?;
+			f.sync_all()?;
+		}
+		fs::rename(&tmp_filename, &filename)?;
+		// Fsync the parent directory on Unix.
+		#[cfg(not(target_os = "windows"))]
+		{
+			let path_str = filename.clone();
+			let path = Path::new(&path_str).parent().unwrap();
+			let dir_file = fs::OpenOptions::new().read(true).open(path)?;
+			unsafe { libc::fsync(dir_file.as_raw_fd()); }
+		}
+		Ok(())
+	}
+
+	#[cfg(test)]
+	fn load_channel_data<ChanSigner: ChannelKeys + Readable + Writeable>(&self) ->
+		Result<HashMap<OutPoint, ChannelMonitor<ChanSigner>>, ChannelMonitorUpdateErr> {
+		if let Err(_) = fs::create_dir_all(&self.path_to_channel_data) {
+			return Err(ChannelMonitorUpdateErr::TemporaryFailure);
+		}
+		let mut res = HashMap::new();
+		for file_option in fs::read_dir(&self.path_to_channel_data).unwrap() {
+			let mut loaded = false;
+			let file = file_option.unwrap();
+			if let Some(filename) = file.file_name().to_str() {
+				if filename.is_ascii() && filename.len() > 65 {
+					if let Ok(txid) = Txid::from_hex(filename.split_at(64).0) {
+						if let Ok(index) = filename.split_at(65).1.split('.').next().unwrap().parse() {
+							if let Ok(contents) = fs::read(&file.path()) {
+								if let Ok((_, loaded_monitor)) = <(BlockHash, ChannelMonitor<ChanSigner>)>::read(&mut Cursor::new(&contents)) {
+									res.insert(OutPoint { txid, index }, loaded_monitor);
+									loaded = true;
+								}
+							}
+						}
+					}
+				}
+			}
+			if !loaded {
+				println!("WARNING: Failed to read one of the channel monitor storage files! Check perms!");
+			}
+		}
+		Ok(res)
+	}
+}
+
+impl<ChanSigner: ChannelKeys + Readable + Writeable + Send + Sync> channelmonitor::Persist<ChanSigner> for FilesystemPersister {
+	fn persist_new_channel(&self, funding_txo: OutPoint, monitor: &ChannelMonitor<ChanSigner>) -> Result<(), ChannelMonitorUpdateErr> {
+		self.write_channel_data(funding_txo, monitor)
+		  .map_err(|_| ChannelMonitorUpdateErr::TemporaryFailure)
```

Comment:
I think these need to be permanent instead - there's no real indication that we'll be able to fix it in the future, and in the mean time we should just go to chain and try to get our money. It will be clearer once we move to better naming - a failed I/O op isn't an "in progress" persist, its a failed one :).

Reply:
Going onchain is really tie to how much backups you have. Loosing one of them is okay and shouldn't be a reason to close all your channels. How do you interpret persistence is really dependent is function of your setup and risk tolerance and it should be clear that is a user policy thing, even if we may give advices/default settings.